src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal class MsQuicStream : TransportConnection
+    {
+        private Task _processingTask;
+        private QuicStream _stream;
+        private MsQuicConnection _connection;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+
+        public MsQuicStream(MsQuicConnection connection, QUIC_STREAM_OPEN_FLAG flags)
+        {
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, 1024 * 1024, 1024 * 1024 / 2, useSynchronizationContext: false);
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, 1024 * 64, 1024 * 64 / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+            if (flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL))
+            {
+                Features.Set<IUnidirectionalStreamFeature>(new UnidirectionalStreamFeature());
+            }
+
+            // Add a fake TLS connection feature becuase quic is always secure
+            var feature = new FakeTlsConnectionFeature();
```

Comment:
Shouldn't this be real and expose the actual TLS information?

Reply:
I'll leave a TODO/file an issue for client certs.

Reply:
You should remove FakeTlsConnectionFeature, ITlsConnectionFeature is optional.